compiler/ballerina-compiler-api/src/main/java/org/ballerina/compiler/impl/TypesFactory.java

```diff
@@ -63,68 +65,134 @@ public static BallerinaTypeDescriptor getTypeDescriptor(BType bType) {
-                return new BallerinaBuiltinTypeDescriptor(moduleID, name, bType);
+                return new BallerinaSimpleTypeDescriptor(moduleID, bType);
+        }
+    }
+
+    private static boolean isTypeReference(BType bType, boolean rawTypeOnly) {
+        if (rawTypeOnly || bType.tsymbol == null) {
+            return false;
+        }
+
+        if ((bType.tsymbol.flags & Flags.ANONYMOUS) == Flags.ANONYMOUS) {
+            return false;
         }
 
-        if (!rawTypeOnly && bType.tsymbol != null && ((bType.tsymbol.flags & Flags.ANONYMOUS) != Flags.ANONYMOUS) &&
-                !bType.tsymbol.getName().getValue().isEmpty()) {
-            typeDescriptorImpl =
-                    new BallerinaTypeReferenceTypeDescriptor(moduleID, bType, bType.tsymbol.getName().getValue());
+        return !bType.tsymbol.getName().getValue().isEmpty();
+    }
+
+    public static TypeDescKind getTypeDescKind(TypeKind bTypeKind) {
+        switch (bTypeKind) {
+            case ANY:
+                return TypeDescKind.ANY;
+            case ANYDATA:
+                return TypeDescKind.ANYDATA;
+            case ARRAY:
+                return TypeDescKind.ARRAY;
+            case BOOLEAN:
+                return TypeDescKind.BOOLEAN;
+            case BYTE:
+                return TypeDescKind.BYTE;
+            case DECIMAL:
+                return TypeDescKind.DECIMAL;
+            case FLOAT:
+                return TypeDescKind.FLOAT;
+            case HANDLE:
+                return TypeDescKind.HANDLE;
+            case INT:
+                return TypeDescKind.INT;
+            case NEVER:
+                return TypeDescKind.NEVER;
+            case NIL:
+                return TypeDescKind.NIL;
+            case STRING:
+                return TypeDescKind.STRING;
+            case JSON:
+                return TypeDescKind.JSON;
+            case XML:
+                return TypeDescKind.XML;
+            case FUNCTION:
+                return TypeDescKind.FUNCTION;
+            case FUTURE:
+                return TypeDescKind.FUTURE;
+            case MAP:
+                return TypeDescKind.MAP;
+            case OBJECT:
+                return TypeDescKind.OBJECT;
+            case STREAM:
+                return TypeDescKind.STREAM;
+            case TUPLE:
+                return TypeDescKind.TUPLE;
+            case TYPEDESC:
+                return TypeDescKind.TYPEDESC;
+            case UNION:
+                return TypeDescKind.UNION;
+            case INTERSECTION:
+                return TypeDescKind.INTERSECTION;
+            case ERROR:
+                return TypeDescKind.ERROR;
+            case ANNOTATION:
+            case BLOB:
+            case CHANNEL:
+            case CONNECTOR:
+            case ENDPOINT:
+            case FINITE:
+            case NONE:
+            case OTHER:
+            case PACKAGE:
+            case READONLY:
+            case SERVICE:
+            case TABLE:
```

Comment:
Shouldn't these two have corresponding typedescs?

Reply:
Probably yes. I only refactored the existing ones. We need to add all the missing types.

Reply:
Ack.